### PR TITLE
Gather the lm_head during evaluation in the Liger distillation trainers

### PR DESCRIPTION
**Pull request by DaoyuanLi2816**
## What does this PR do?

`training_step` in the Liger distillation trainers wraps the step in `_get_liger_zero3_lm_head_gather_ctx`, because the fused JSD projects through the `lm_head` and that projection needs the head gathered under Liger + ZeRO-3. Evaluation goes through `prediction_step`, which does the same projection but without the gather, so eval on a ZeRO-3 run sees a sharded `lm_head`.

This adds the same gather to the evaluation path:

- `SDFTTrainer.prediction_step` and `SDPOTrainer.prediction_step` — both already had the gather in `training_step` only.
- `IWOPDTrainer` — had no `prediction_step` override at all, so it now gets one mirroring the other two.
- `GOLDTrainer` — has its own `prediction_step` override that calls `compute_loss` directly rather than `super()`, so the gather wraps that call.

It also makes the Liger path return the student outputs when `return_outputs=True`, in both `IWOPDTrainer.compute_loss` and `GOLDTrainer.compute_loss`. The non-Liger path a few lines below already returns them; the Liger path returned `(loss, None)` in IWOPD and never bound the name it returns in GOLD. `Trainer.prediction_step` slices `outputs[1:]` when `prediction_loss_only=False`, which works on neither.

## Correction to an earlier revision of this description

An earlier revision of this description claimed `GOLDTrainer` had changed on `main` and no longer needed anything here. That was wrong. Cursor Bugbot flagged it on this PR and was right on both counts. Checked against current `main` (`2396dfe`):

- `GOLDTrainer.prediction_step` (`gold_trainer.py` L2761) calls `self.compute_loss` with no gather, while `training_step` (L2786) has one. `_get_liger_zero3_lm_head_gather_ctx` is defined 25 lines above the override and used only by `training_step`.
- In `compute_loss`, the Liger branch binds `student_outputs`, deletes it at L2525, and never binds `outputs_student`, so the shared `return (loss, outputs_student) if return_outputs else loss` at L2631 raises `UnboundLocalError` on that path.

`GOLDTrainer` subclasses `SFTTrainer` and declares exactly one `prediction_step`, so nothing else covers it. Both fixes are included here now.

The `DistillationTrainer` half of that earlier note still holds, re-checked at the same commit: it routes the whole loss through `_forward_redirection` into a unified `_compute_loss`, has no `_get_liger_zero3_lm_head_gather_ctx` at all, and its `(loss, None)` is deliberate because the chunked path never materializes student logits.

The branch was rebuilt on current `main` rather than merged, since the original diff no longer applied.

## Tests

Each new test counts calls to `_get_liger_zero3_lm_head_gather_ctx` during `trainer.evaluate()`. All four fail without the corresponding source change (0 calls) and pass with it.

- `pytest tests/experimental/test_sdft_trainer.py tests/experimental/test_sdpo_trainer.py tests/experimental/test_iw_opd_trainer.py -k prediction_step_gathers`: 3 passed
- `pytest tests/experimental/test_sdft_trainer.py tests/experimental/test_sdpo_trainer.py tests/experimental/test_iw_opd_trainer.py`: 77 passed, 3 skipped
- `pytest tests/experimental/test_gold_trainer.py -k prediction_step_gathers`: 1 passed; reverting only `gold_trainer.py` makes it fail, so the test is anchored to the fix
- `pytest tests/experimental/test_gold_trainer.py`: 67 passed, 4 skipped
- `ruff check` and `ruff format --check` on the changed files

Two limits worth stating rather than leaving implied:

- The `return_outputs=True` fixes have no dedicated test in either trainer. Reaching that branch needs the Liger kernel path, which the CPU test environment here does not exercise; the change is argued from control flow, where the branch plainly never binds the name it returns.
- The ZeRO-3 branch of the gather itself still needs a multi-GPU DeepSpeed run. The tests cover that the context is entered on the eval path, not the DeepSpeed behaviour inside it.

## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

## AI writing disclosure

We welcome the use of AI tools to help with contributions. For transparency and to help us improve our review process, please indicate the level of AI involvement in this PR.

- [ ] No AI usage: the PR was written entirely by a human.
- [ ] AI-assisted: some parts were suggested or improved by AI, but the PR was written and reviewed by a human.
- [x] AI-generated: the PR was mostly or fully generated by an AI tool.

## Who can review?

@qgallouedec

---

> [!NOTE]
> **Medium Risk**
> Changes eval loss computation for Liger + DeepSpeed ZeRO-3 runs and touches multiple experimental trainers; behavior is intentionally aligned with training but affects reported eval metrics and any code relying on Liger `return_outputs`.
> 
> **Overview**
> Evaluation for **Liger** distillation trainers now wraps `prediction_step` in `_get_liger_zero3_lm_head_gather_ctx`, matching `training_step`, so fused JSD can project through a gathered `lm_head` under ZeRO-3 instead of a sharded one.
> 
> **SDFTTrainer**, **SDPOTrainer**, and **GOLDTrainer** add the gather around `compute_loss` in their existing `prediction_step` overrides. **GOLDTrainer** also sets `outputs_student` on the Liger branch so `return_outputs=True` does not reference an unbound name.
> 
> **IWOPDTrainer** adds a `prediction_step` override with the same gather, routes Liger `compute_loss` through `_compute_liger_loss` with `return_outputs`, and returns `(loss, student_outputs)` from the Liger path when outputs are requested (aligned with the non-Liger path).
> 
> Regression tests in the GOLD, IW-OPD, SDFT, and SDPO test modules assert `_get_liger_zero3_lm_head_gather_ctx` is invoked during `trainer.evaluate()`.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 558cabb3d055f47c8c4494e560894d8539fc328e. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by qgallouedec** [comment]
Thanks @DaoyuanLi2816, I reproduced both bugs on `main` and the fix is minimal and correct: `training_step` already gathers the sharded `lm_head` via `_get_liger_zero3_lm_head_gather_ctx` while `prediction_step` didn't, and the liger branch's `student_outputs`/`outputs_student` name mismatch does raise `UnboundLocalError` on `return_outputs=True`.

My one ask is consistency across the liger-trainer family: exactly four trainers use that gather context (sdpo, sdft, gold, distillation) and all four have the same eval-path gap. This PR fixes gold and #6384 covers sdpo/sdft, but `distillation_trainer.py` has no `prediction_step` override and is fixed by neither. Could you fold the distillation eval-gather fix into this effort and coordinate with #6384 so the whole family lands the gather together, rather than merging gold alone? (The `student_outputs → outputs_student` rename is a nice consistency win with the ULD/standard branches — keep it.)

**Comment by DaoyuanLi2816** [comment]
Consolidated the Distillation and IWOPD output fixes from #6393 into this branch. The five-trainer evaluation coverage now passes (6 tests; 2 hardware-dependent tests skipped).

**Comment by DaoyuanLi2816** [comment]
This had gone stale and conflicted, so I rebuilt it on current `main` and rescoped it. Summary of what changed since your review:

`GOLDTrainer` and `DistillationTrainer` are both out — they no longer need anything. `GOLDTrainer` now gathers inside its own `prediction_step` and returns `(loss, outputs_student)`. `DistillationTrainer` was rewritten to route the loss through `_forward_redirection` into a unified `_compute_loss`; it has no `_get_liger_zero3_lm_head_gather_ctx` any more, and its `(loss, None)` looks deliberate given the chunked path never materializes student logits, so I left it alone.

What is left is the same bug in the three trainers that still have it: `SDFTTrainer` and `SDPOTrainer` gather in `training_step` but not in `prediction_step`, and `IWOPDTrainer` has no `prediction_step` override at all despite gathering in `training_step`. `IWOPDTrainer` also still returns `(loss, None)` on the Liger path while its own non-Liger path returns `(loss, student_outputs)`.

The three new tests count calls to `_get_liger_zero3_lm_head_gather_ctx` during `evaluate()`; I checked they fail on `main` and pass here. Full run of the three affected test files: 77 passed, 3 skipped.

The branch was rebuilt rather than merged because the original diff no longer applied, so this is a force-push. Sorry for the churn.

File: trl/experimental/sdpo/sdpo_trainer.py [hunk 1]
```diff
@@ -822,7 +822,10 @@ def prediction_step(self, model, inputs, prediction_loss_only, ignore_keys=None)
             inputs = self._prepare_inputs(inputs)
         with torch.no_grad():
             with self.compute_loss_context_manager():
-                loss = self.compute_loss(model, inputs)
+                # Same gather as `training_step`: the fused JSD projects through the lm_head, which stays sharded
+                # under Liger + ZeRO-3 unless it is gathered first.
+                with self._get_liger_zero3_lm_head_gather_ctx(model):
+                    loss = self.compute_loss(model, inputs)
```
[the remark this answers is not in the document]
Comment (DaoyuanLi2816) [reply trl/experimental/sdpo/sdpo_trainer.py:828]: Confirmed, and fixed in 558cabb — thanks, this was a real miss on my side.

`GOLDTrainer.prediction_step` was calling `compute_loss` with no gather while `training_step` had one, and the PR description wrongly said `main` had already handled it. I've corrected that section of the description too.

The gather now wraps that call, with a regression test that counts context entries during `evaluate()`: it fails with only the test applied and passes with the source change. `pytest tests/experimental/test_gold_trainer.py`: 67 passed, 4 skipped.

File: trl/experimental/iw_opd/iw_opd_trainer.py [hunk 2]
```diff
@@ -1752,7 +1753,7 @@ def _compute_liger_loss(self, model, inputs, num_items_in_batch=None):
             loss = loss * num_valid_local / num_items_in_batch
 
         del student_hidden, teacher_hidden, true_labels
-        return loss
+        return (loss, student_outputs) if return_outputs else loss
```
[the remark this answers is not in the document]
Comment (DaoyuanLi2816) [reply trl/experimental/iw_opd/iw_opd_trainer.py:1756]: Confirmed, and fixed in 558cabb.

The Liger branch bound `student_outputs`, deleted it at L2525, and never bound `outputs_student`, so the shared `return (loss, outputs_student) if return_outputs else loss` at L2631 raises `UnboundLocalError` on that path. It now binds before the delete, mirroring the `IWOPDTrainer` change already in this PR.

One limit worth stating: there's no dedicated test for this half, in either trainer. Reaching the branch needs the Liger kernel path, which the CPU environment here doesn't exercise, so it rests on the control flow rather than on coverage.

**Comment by DaoyuanLi2816** [comment]
@qgallouedec Could you revisit the evaluation-gather changes on the current head (558cabb3) when you have a chance? This is still awaiting review after the earlier updates. The latest tests/docs runs are labeled failed but contain zero jobs, so they do not provide a test result; could those workflows be approved or rerun for a fresh CI signal?

**Comment by DaoyuanLi2816** [comment]
Closing as superseded by #7301, which removed the fused Liger JSD paths this PR fixes in GOLD, IW-OPD, SDFT and SDPO.
